### PR TITLE
Transform coords without dense coord

### LINKED ISSUE #2171: `transform_coords` does not work with event coords if no bin coord present
This
```.py
import scipp as sc
events = sc.DataArray(sc.arange('event', 5.0),
                      coords={'per_event': sc.arange('event', 5),
                              'per_bin': sc.array(dims=['event'], values=[0, 0, 0, 1, 1])})
da = sc.bin(events, groups=[sc.array(dims=['per_bin'], values=[0, 1])])
def x(per_event):
    return per_event
graph = {'x': x}
sc.transform_coords(da, 'x', graph)
```
raises the following during `finalize`
```
.../scipp/coords.py in _add_coord(self, name)
    116         if name in self.obj.meta:
    117             return _produce_coord(self.obj, name)
--> 118         if isinstance(self._graph[name], str):
    119             self._aliases.append(name)
    120             out = self._get_coord(self._graph[name])

.../scipp/coords.py in __getitem__(self, name)
     28 
     29     def __getitem__(self, name):
---> 30         return self._graph[name]
     31 
     32     def __contains__(self, name):

KeyError: 'per_event'
```

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #2171.

### CODE REVIEW

File: python/src/scipp/coords.py [hunk 1]
```diff
@@ -98,60 +120,67 @@ def __init__(self, obj, *, graph, outputs):
...
+            have_all_dense_inputs = all([v[0] is not None for v in args.values()])
+            if have_all_dense_inputs:
+                dense_args = {k: v[0] for k, v in args.items()}
+                out = func(**dense_args)
+                if not isinstance(out, dict):
+                    out = {name: out}
+            else:
+                out = {}
+            have_event_inputs = any([v[1] is not None for v in args.values()])
+            if self.obj.bins is not None and have_event_inputs:
```
Comment (jl-wynen) [inline python/src/scipp/coords.py:143]: Is the `self.obj.bins is not None` check required? If `have_event_inputs` is `True`, does that not imply that the object has bins?

**Review by jl-wynen — COMMENTED** [review]
Looks fine, but seems quite complicated because of all the nested logic.

File: python/src/scipp/coords.py (hunk 1 above)
Comment (SimonHeybrock) [reply python/src/scipp/coords.py:143]: 👍

File: python/src/scipp/coords.py [hunk 2]
```diff
@@ -57,27 +57,49 @@ def show(self, size=None, simplified=False):
...
 def _store_coord(obj, name, coord):
-    obj.coords[name] = coord
+    dense_coord, event_coord = coord
+    if dense_coord is not None:
+        obj.coords[name] = dense_coord
     if name in obj.attrs:
         # If name is both an input and output to a function,
         # the input handling made it an attr, but since it is
         # an output, we want to store it as a coord (and only as a coord).
         del obj.attrs[name]
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply python/src/scipp/coords.py:98]: Good catch. Fixed in (now renamed) helper function.

**Comment by SimonHeybrock** [comment]
> Looks fine, but seems quite complicated because of all the nested logic.

Yes, unfortunately. Is handling this using `tuple(dense, event)` the wrong approach, or can you think of another culprit?

**Comment by jl-wynen** [comment]
> Is handling this using tuple(dense, event) the wrong approach, or can you think of another culprit?

I am not sure. We need to extract both and need to keep track of whether they apply to the outer array or the events.
I had some ideas about the general structre reading through the code. But without fleshing them out, I don't know if they would simplify anything. Basically, instead of walking the graph and computing coords at the same time, find a path through the graph first and construct a smaller one. That one, I think could be turned into a linear sequence. Then, traverse that sequence and apply conversions as you go. This would not address the event-dense issue but would maybe simplify the overall structure.
Maybe we can take a look at a later point in time if this is relevant enough.

**Pull request merged**